text/2018-10-25-tikv-client-rust.md

```diff
@@ -0,0 +1,395 @@
+
+### Try the Transactional Key-Value API
+
+The Transactional Key-Value API is more complicated than the Raw Key-Value API. Some transaction related concepts are listed as follows.
+
+- Client
+
+    Like the Raw Key-Value client, a Client is client to a TiKV cluster.
+
+- Snapshot
+
+    A Snapshot is the state of a Client at a particular point of time, which provides some readonly methods. The multiple reads of the same Snapshot is guaranteed consistent.
+
+- Transaction
+
+    Like the Transaction in SQL, a Transaction symbolizes a series of read and write operations performed within the Client. Internally, a Transaction consists of a Snapshot for reads, and a buffer for all writes. The default isolation level of a Transaction is Snapshot Isolation.
+
+To use the Transactional Key-Value API, take the following steps:
+
+1. Create an instance of Config to specify endpoints of PD (Placement Driver) and optional security config
+
+    ```rust
+        let config = Config::new(vec!["127.0.0.1:2379"]);
+    ```
+
+2. Create a Transactional Key-Value client.
+
+    ```rust
+        let client = TxnClient::new(&config);
+    ```
+
+3. (Optional) Modify data using a Transaction.
+
+    The lifecycle of a Transaction is: _begin → {get, set, delete, scan} → {commit, rollback}_.
+
+4. Call the Transactional Key-Value API's methods to access the data on TiKV. The Transactional Key-Value API contains the following methods:
+
+    ```rust
+        fn begin(&self) -> KvFuture<Transaction>;
+
+        fn get<K>(&self, key: K) -> KvFuture<Value>
+        where
+            K: AsRef<Key>;
+
+        fn set<P>(&mut self, pair: P) -> KvFuture<()>
+        where
+            P: Into<KvPair>;
+
+        fn delete<K>(&mut self, key: K) -> KvFuture<()>
+        where
+            K: AsRef<Key>;
+
+        fn seek<K>(&self, key: K) -> KvFuture<Scanner>
+        where
+            K: AsRef<Key>;
+
+        fn commit(self) -> KvFuture<()>;
+
+        fn rollback(self) -> KvFuture<()>;
+    ```
+
+### Usage example of the Transactional Key-Value API
+
+```rust
+    extern crate futures;
+    extern crate tikv_client;
+
+    use futures::{Async, Future, Stream};
+    use tikv_client::transaction::{Client, Mutator, Retriever, TxnClient};
+    use tikv_client::*;
+
+    fn puts<P, I>(client: &TxnClient, pairs: P)
+    where
+        P: IntoIterator<Item = I>,
+        I: Into<KvPair>,
+    {
+        let mut txn = client.begin().wait().expect("Could not begin transaction");
+        let _: Vec<()> = pairs
+            .into_iter()
+            .map(Into::into)
+            .map(|p| {
+                txn.set(p).wait().expect("Could not set key value pair");
+            }).collect();
+        txn.commit().wait().expect("Could not commit transaction");
+    }
+
+    fn get(client: &TxnClient, key: &Key) -> Value {
+        let txn = client.begin().wait().expect("Could not begin transaction");
+        txn.get(key).wait().expect("Could not get value")
```

Comment:
What happens to the transaction at this point?

Reply:
I assume that transaction will be dropped before returning to caller. So basically it is trying to read something in a transaction.